pkg/serviceaccount/metadata_test.go

```diff
@@ -0,0 +1,240 @@
+	// TokenEndpoint `json:"token_endpoint"` // "REQUIRED unless only the Implicit Flow is used."
+	JwksURI       string   `json:"jwks_uri"`
+	ResponseTypes []string `json:"response_types_supported"`
+	SigningAlgs   []string `json:"id_token_signing_alg_values_supported"`
+	SubjectTypes  []string `json:"subject_types_supported"`
+}
+
+func TestServeConfiguration(t *testing.T) {
+	s := setupServer(t, "my-fake-issuer", defaultKeys)
+	defer s.Close()
+
+	want := Configuration{
+		Issuer:        "my-fake-issuer",
+		JwksURI:       s.URL + "/serviceaccountkeys/v1/jwks.json",
+		ResponseTypes: []string{"id_token"},
+		SubjectTypes:  []string{"public"},
+		SigningAlgs:   []string{"ES256", "RS256"},
+	}
+
+	reqURL := s.URL + "/.well-known/openid-configuration"
+
+	resp, err := http.Get(reqURL)
+	if resp == nil || err != nil {
+		t.Errorf("Get(%s) = %v, %v want: <response>, %v", reqURL, resp, err, nil)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		t.Errorf("Get(%s) = %v, _ want: %v, _", reqURL, resp.StatusCode, http.StatusOK)
+	}
+
+	if got, want := resp.Header.Get("Content-Type"), "application/json"; got != want {
+		t.Errorf("Get(%s) Content-Type = %q, _ want: %q, _", reqURL, got, want)
+	}
+
+	d := json.NewDecoder(resp.Body)
+	var got Configuration
+	if err := d.Decode(&got); err != nil {
+		t.Errorf("Decode(_) = %v, want: <nil>", err)
+		return // can't evaluate Configuration
+	}
+
+	if !reflect.DeepEqual(got, want) {
+		t.Errorf("received configuration differs: got: %+v\nwant: %+v\n", got, want)
+	}
+}
+
+var serveKeysTests = []struct {
+	Name     string
+	Keys     []interface{}
+	WantKeys []jose.JSONWebKey
+}{
+	{
+		Name: "configured public keys",
+		Keys: []interface{}{
+			getPublicKey(rsaPublicKey),
+			getPublicKey(ecdsaPublicKey),
+		},
+		WantKeys: []jose.JSONWebKey{
+			{Key: &rsa.PublicKey{}},
+			{Key: &ecdsa.PublicKey{}},
+		},
+	},
+	{
+		Name: "only publishes public keys",
+		Keys: []interface{}{
+			getPrivateKey(rsaPrivateKey),
+			getPrivateKey(ecdsaPrivateKey),
+		},
+		WantKeys: []jose.JSONWebKey{
+			{Key: &rsa.PublicKey{}},
+			{Key: &ecdsa.PublicKey{}},
+		},
+	},
+}
+
+func TestServeKeys(t *testing.T) {
+	for _, tt := range serveKeysTests {
+		t.Run(tt.Name, func(t *testing.T) {
+			s := setupServer(t, "my-fake-issuer", tt.Keys)
+			defer s.Close()
+
+			reqURL := s.URL + "/serviceaccountkeys/v1/jwks.json"
+
+			resp, err := http.Get(reqURL)
+			if resp == nil || err != nil {
+				t.Errorf("Get(%s) = %v, %v want: <response>, %v", reqURL, resp, err, nil)
+			}
+			if resp.StatusCode != http.StatusOK {
+				t.Errorf("Get(%s) = %v, _ want: %v, _", reqURL, resp.StatusCode, http.StatusOK)
+			}
+			if got, want := resp.Header.Get("Content-Type"), "application/jwk-set+json"; got != want {
+				t.Errorf("Get(%s) Content-Type = %q, _ want: %q, _", reqURL, got, want)
+			}
+			if resp.Body == nil {
+				t.Errorf("resp.Body = %v, want io.ReadCloser", resp.Body)
+				return // can't evaluate body
+			}
+			defer resp.Body.Close()
```

Comment:
Call this after checking that the error is not nil.

Reply:
Done